### ISSUE TITLE
Image Conversion Occurs After Upload to Cloud Storage, Resulting in Non-Resized Images Being Stored

### ISSUE THREAD

**Issue by rafaelrat** [payloadcms/plugin-cloud-storage#67]
**Description:**
I am encountering an issue where the conversion of images to resized formats is taking place after the images are uploaded to cloud storage. As a result, the stored images are in their original, non-resized format instead of the intended resized format.

**Steps to Reproduce:**
1. Configure the collection to convert image files to resized formats.
2. Upload an image.
3. Observe that the image is stored in its original, non-resized format in cloud storage.

**Collection Configuration:**
```typescript
upload: {
    disableLocalStorage: true,
    formatOptions: {
        format: 'webp',
    },
    mimeTypes: ['image/*']
}
``` 

**Expected Behavior:**
The image resizing should occur before the image is uploaded to cloud storage. This would ensure that the stored images are in the intended resized format as configured in the collection settings.

**Actual Behavior:**
The image resizing appears to be occurring after the upload process, causing the stored images to remain in their original, non-resized format.

**Comment by rafaelrat** [comment]
I've noticed that when using the imageSizes property, the files for each image size are being resized and saved correctly. However, the original image is still being saved without being resized.

```typescript
upload: {
    disableLocalStorage: true,
    formatOptions: {
        format: 'webp',
    },
    imageSizes: [
        {
            name: 'icon',
            formatOptions: {
                format: 'webp'
            }
        }
    ],
    mimeTypes: ['image/*']
}

```

**Comment by sdjnes** [comment]
We are seeing the same issue. Payload reports the compressed size in the UI, but the file that is uploaded to our GCP bucket is the original image and not the compressed one.

**Comment by sdjnes** [comment]
On create and update of collections, Payload itself executes the following code (https://github.com/payloadcms/payload/blob/37fe2df2de41b4a64cbda228e6adad349aefa4c0/src/collections/operations/create.ts#L155C1-L176)

```js
  // /////////////////////////////////////
  // Write files to local storage
  // /////////////////////////////////////


  if (!collectionConfig.upload.disableLocalStorage) {
    await uploadFiles(payload, filesToUpload, req.t);
  }


  // /////////////////////////////////////
  // beforeChange - Collection
  // /////////////////////////////////////


  await collectionConfig.hooks.beforeChange.reduce(async (priorHook, hook) => {
    await priorHook;


    data = (await hook({
      data,
      req,
      operation: 'create',
      context: req.context,
    })) || data;
  }, Promise.resolve());
```

This plugin registers `beforeChange` and `afterDelete` hooks (see: https://github.com/payloadcms/plugin-cloud-storage/blob/3d5858ca6d45d73c22e9bb97c2b443758d93400b/src/plugin.ts#L79-L89)

From the above Payload snippet, the `beforeChange` hooks are run after the `uploadFiles` call which explains why the compressed/resized images are not being uploaded. However, earlier in the same function, the `beforeOperation` hooks are called ([source](https://github.com/payloadcms/payload/blob/37fe2df2de41b4a64cbda228e6adad349aefa4c0/src/collections/operations/create.ts#L54-L62)).

One slightly confusing thing for me is whether or not a series of `beforeOperation` hooks and a `beforeChange` hook should be equivalent such as:

`beforeOperation(args, 'create') && beforeOperation(args, 'update') === beforeChange(args)`

The docs seem to elude that they _are_ equivalent "beforeChange hooks will run within create and update operations" ([source](https://payloadcms.com/docs/hooks/collections#beforechange)).

Therefore I'm not sure where the change should be made. It could be possible to make the change in this plugin to instead register two `beforeOperation` hooks for `change` and `update`. However, I think the smarter solution is for Payload itself to be changed so that `beforeOperation` and `beforeChange` (and possibly all other `before...` hooks) are executed at the same time.

**Comment by denolfe** [comment]
Thank you for the reports. We're looking into this one.

**Comment by denolfe** [comment]
This was fixed by https://github.com/payloadcms/payload/pull/3883 and https://github.com/payloadcms/payload/pull/3883 and https://github.com/payloadcms/payload/pull/3996. Released in 2.0.15. The fix was related to payload core.
